k-distribution/k-tutorial/1_basic/22_proofs/README.md

```diff
@@ -1 +1,448 @@
+        STORE [ $a <- A:Int ] [ $b <- B:Int ] [ $c <- B:Int ]
+    }
+  #And
+    {
+      true
+    #Equals
+      ?C ==Int A orBool ?C ==Int B
+    }
+  )
+#And
+  <generatedTop>
+    <k>
+      _DotVar1
+    </k>
+    <store>
+      STORE [ $a <- A:Int ] [ $b <- B:Int ] [ $c <- B:Int ]
+    </store>
+    <funcs>
+      _Gen3
+    </funcs>
+    <stack>
+      _Gen5
+    </stack>
+  </generatedTop>
+#And
+  {
+    true
+  #Equals
+    A <Int B
+  }
+```
+
+This output starts with a message telling us at which point the proof failed,
+followed by the final state, which consists of three parts: some negative
+Matching Logic (ML) constraints, the final configuration (`<generatedTop> ...
+</generatedTop>`), and some positive ML constraints. Generally speaking,
+these positive and the negative constraints could arise from various sources,
+such as (but not limited to) branches taken by the execution
+(e.g. `{ true #Equals A <Int B }` or `#Not ( { true #Equals A <Int B>} ) `),
+or `ensures` constraints.
+
+First, we examine the message:
+
+```
+(InfoReachability) while checking the implication:
+The configuration's term unifies with the destination's term,
+but the implication check between the conditions has failed.
+```
+
+which tells us that the *structure* of the final configuration is as expected,
+but that some of the associated constraints cannot be proven. We next look at
+the final configuration, in which the relevant item is the `<store> ...
+</store>` cell, because it is the only one that we are reasoning about. By
+inspecting its contents:
+
+```
+STORE [ $a <- A:Int ] [ $b <- B:Int ] [ $c <- B:Int ]
+```
+
+we see that we should be within the constraints of the `ensures`, since the
+value of `$c` in the store equals `B` in this branch. We next examine the
+negative and positive constraints of the output and, more often than not, the
+goal is to instruct K how to use the information from the final configuration
+and the positive constraints to falsify one of the negative constraints. This
+is done through *simplifications*.
+
+So, the positive constraint that we have is
+
+```
+{ true #Equals A <Int B }
+```
+
+meaning that `A <Int B` holds. Given the analysed program, this tells us that
+we are in the `then` branch of the `if`. The negative constraint is
+
+```
+  #Not (
+    #Exists ?C . {
+        STORE [ $a <- A:Int ] [ $b <- B:Int ] [ $c <- ?C:Int ]
+      #Equals
+        STORE [ $a <- A:Int ] [ $b <- B:Int ] [ $c <- B:Int ]
+    }
+  #And
+    { true #Equals ?C ==Int A orBool ?C ==Int B }
+  )
+```
+
+and we observe, from the first equality, that the existential `?C` should be
+instantiated with `B`. This would make both branches of the `#And` true,
+falsifying the outside `#Not`. We just need to show `K` how to conclude that
+`?C ==Int B`. We do so by introducing the following simplification into the
+`VERIFICATION` module:
+
+```
+rule { M:Map [ K <- V ] #Equals M [ K <- V' ] } => { V #Equals V' } [simplification, anywhere]
```

Comment:
The `anywhere` attribute should be removed since `[simplification, anywhere]` is now forbidden.

```suggestion
rule { M:Map [ K <- V ] #Equals M [ K <- V' ] } => { V #Equals V' } [simplification]
```